proposals/0000-complete-set-signatures.rst

```diff
@@ -0,0 +1,483 @@
+
+Proposed Change Specification
+-----------------------------
+
+Syntax
+======
+
+Extend Syntax so that we are able to write the example from the previous section
+
+::
+
+ {-# COMPLETE (:<), Empty :: IsList l => l a #-} -- Context is allowed
+ {-# COMPLETE (:<) :: Stream a #-}               -- TyCon App is allowed
+ {-# COMPLETE (:<), Empty :: String #-}          -- Even type synonyms
+
+Since there is no formal grammar for COMPLETE pragmas, here's how I propose to
+change the happy grammar from
+
+::
+
+ sigdecl
+   : ...
+   | '{-# COMPLETE' con_list opt_tyconsig  '#-}'
+ opt_tyconsig
+   : {- -}
+   | '::' gtycon
+ gtycon
+   : ntgtycon
+   | '(' ')'
+   | '(#' '#)'
+
+to
+
+::
+
+ sigdecl
+   : ...
+   | '{-# COMPLETE' con_list opt_restysig  '#-}'
+ opt_restysig
+   : {- -}
+   | '::' ctype
+ atype
+   : ntgtycon
+   | '(' ')'
+   | '(#' '#)'
+   | ... many more ...
+
+Where a ``ctype`` is a "for-all type", according to GHC's happy parser. It's the
+same grammatical sort that constitutes the RHS of a ``::`` in a type signature
+and since it constrains the result type of the relevant ConLikes, we call
+it a "result type signature", often referring just to the type it carries.
+
+Note that this change accepts strictly more syntax, because
+
+- ``atype`` accepts a superset of ``gtycon``
+- ``atype`` can be derived from ``ctype`` (via ``type``, ``btype``, ``infixtype``, ``ftype``)
+
+Typing
+======
+
+The (type carried by the) result type signature must have kind ``TYPE r``, for
+any runtime-representation ``r``. Any free type variables are implicitly bound
+by top-level ``forall``, just like it happens for function type signatures.
+
+Examples for well-typed result type signatures:
+
+::
+
+ [a]
+ IsInfinite l => l a
+ forall l a. IsInfinite l => l a  -- equivalent to the former
+ String
+ (a ~ Int, Semigroup a) => a
+ MPTC a b => a
+ Int#
+
+Examples for invalid result type signatures:
+
+::
+
+ Stream
+ (->)
+ "Symbol"
+ Eq Int
+
+Note that after type-checking
+
+- We *accept* some of the previously well-typed syntax, like ``Int`` and other
+  nullary data type constructors.
+- We *reject* some of the previously well-typed syntax, like ``Stream`` and other
+  non-nullary data type constructors.
+- We *accept* new syntax, like ``IsInfinite l => l a``.
+
+Pattern match checking
+======================
+
+A COMPLETE set with a result type signature ``{-# COMPLETE cls :: sig_ty
+#-}`` is to be treated the same as one without, with one exception: When
+the COMPLETE set is supposedly *covered* by a set of patterns in a pattern
+match, we
+
+1. Take the result type of the pattern match, ``ty``.
+2. Check whether ``sig_ty`` subsumes ``ty``, as per the usual subsumption
+   rules of GHC.
+   If that is the case, then the COMPLETE set is *covered* by the pattern match.
+   Otherwise, the COMPLETE set is *not covered* by the pattern match.
+
+If *any* COMPLETE set is covered by a pattern match, then the pattern match is
```

Comment:
This isn't precise enough.  We have two things in play
1. Which COMPLETE sets to consider as "candidates"... for that we do the subsumption test thing.
2. Of those candidates, does the pattern match mention all of the constructors in the candidate?

For step 2, there's a GADT thing to consider; e.g. 
```
f :: Fin Z -> blah
f FZ = ...
```
is exhaustive, even with `{-# COMPLETE FZ, FS #-}`